lib/server/services/qualityControlFlag/QcFlagService.js

```diff
@@ -883,6 +528,138 @@ class QcFlagService {
             .orderBy('createdAt', 'DESC', 'verifications');
     }
 
+    /**
+     * Validate QC flag timestamps
+     * If null timestamp was provided, given timestamp is replaced by run's startTime or endTime
+     * @param {Partial<Period>} timestamps QC flag timestamps
+     * @param {Run} targetRun run which for QC flag is to be set
+     * @return {{from: (number|null), to: (number|null)}} prepared timestamps
+     * @throws {BadParameterError}
+     */
+    _prepareQcFlagPeriod(timestamps, targetRun) {
+        const { qcTimeStart, qcTimeEnd } = targetRun;
+        const runStart = qcTimeStart?.getTime();
+        const runEnd = qcTimeEnd?.getTime();
+
+        const { from, to } = timestamps;
+
+        if (from && to && from >= to) {
+            throw new BadParameterError('Parameter "to" timestamp must be greater than "from" timestamp');
+        }
+
+        const isFromOutOfRange = from && (runStart && from < runStart || runEnd && runEnd <= from);
+        const isToOutOfRange = to && (runStart && to <= runStart || runEnd && runEnd < to);
+        if (isFromOutOfRange || isToOutOfRange) {
+            throw new BadParameterError(`Given QC flag period (${from}, ${to}) is out of run (${runStart}, ${runEnd}) period`);
+        }
+
+        return { from, to };
+    }
+
+    /**
+     * Remove a time segment from a list of QC flags effective periods
+     *
+     * @param {Partial<Period>} eraseWindow time segment that should be removed from given periods
+     * @param {SequelizeQcFlagEffectivePeriod[]} periods effective periods from which time window should be erased
+     * @return {Promise<void>} resolve once all periods are updated
+     */
+    async _removeEffectivePeriodsAndPeriodIntersection(eraseWindow, periods) {
+        /*
+         * Example of what is named `before` and `after` an erase window:
+         *
+         * ------------------------------[ erase window ]----------------------------> time
+         * [ period BEFORE erase window ]                [ period AFTER erase window ]
+         *
+         * Using this flag for example:
+         * ---------------------[        QC flag period         ]--------------------
+         *
+         * The corresponding sub-periods will be:
+         * ---------------------[       ]-------------------------------------------- => sub-period BEFORE erase window
+         * ----------------------------------------------[      ]-------------------- => sub-period AFTER erase window
+         *
+         * The following flag would have no sub-period after:
+         * ----------------[ QC flag period ]----------------------------------------
```

Comment:
Analogously, for the following QC flag just right-hand time segment will be erased

Reply:
Rephrased